src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs

```diff
@@ -2,213 +2,1555 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Generic;
-using System.Collections.Immutable;
 using System.Diagnostics;
-using Microsoft.CodeAnalysis.Differencing;
-using Microsoft.CodeAnalysis.Text;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
 namespace Microsoft.CodeAnalysis.CSharp.EditAndContinue
 {
-    internal abstract class SyntaxComparer : TreeComparer<SyntaxNode>
+    internal sealed class SyntaxComparer : AbstractSyntaxComparer
     {
-        internal const int IgnoredNode = -1;
+        internal static readonly SyntaxComparer TopLevel = new();
+        internal static readonly SyntaxComparer Statement = new(compareStatementSyntax: true);
+
+        private readonly SyntaxNode? _oldRoot;
+        private readonly SyntaxNode? _newRoot;
+        private readonly IEnumerable<SyntaxNode>? _oldRootChildren;
+        private readonly IEnumerable<SyntaxNode>? _newRootChildren;
+        private readonly bool _compareStatementSyntax;
```

Comment:
>private readonly bool _compareStatementSyntax; [](start = 8, length = 46)

Would be good to add a comment explaining what "statement mode" means - e.g. top-level treats method bodies as opaque vs. statement-level descends into statements and expressions.